File: app/src/main/java/com/imperium/academio/ui/fragment/AttendanceStudent.java
package com.imperium.academio.ui.fragment;

import android.app.Activity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;
import androidx.core.content.res.ResourcesCompat;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.imperium.academio.CustomUtil;
import com.imperium.academio.R;
import com.imperium.academio.databinding.FragmentAttendanceStudentBinding;
import com.imperium.academio.databinding.TemplateAttendanceDateBinding;
import com.imperium.academio.fireclass.AttendanceHelperClass;
import com.imperium.academio.fireclass.ClassHelperClass;

import java.text.DateFormatSymbols;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class AttendanceStudent extends Fragment {
    FragmentAttendanceStudentBinding binding;
    DatabaseReference selectedClass;
    String classId;
    String userId;
    List<Long> absentList;
    List<Long> sessionList;

    public AttendanceStudent() {
        // Required empty public constructor
    }

    public static AttendanceStudent newInstance(Bundle args) {
        AttendanceStudent fragment = new AttendanceStudent();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle bundle = getArguments();
        if (bundle != null) {
            classId = bundle.getString("classId");
            userId = bundle.getString("userId");
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        binding = DataBindingUtil.inflate(
                inflater, R.layout.fragment_attendance_student, container, false
        );
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Activity activity = getActivity();
        LayoutInflater inflater = getLayoutInflater();
        if (activity == null || classId == null || userId == null) return;

        selectedClass = FirebaseDatabase.getInstance().getReference("class/" + classId);
        absentList = new ArrayList<>();
        sessionList = new ArrayList<>();
        fetchAttendanceSheet();

        // Temporarily fill progress bar
        setProgress(0, 0);

        // Table creation
        TableLayout table = binding.attendanceStudentTable;
        table.setStretchAllColumns(true);
        table.bringToFront();
        String[] shortWeekDays = DateFormatSymbols.getInstance().getShortWeekdays();
        for (int i = 0; i < 7; i++) {
            // For row
            TableRow tr = new TableRow(activity);

            for (int j = 0; j < 7; j++) {
                // For Element
                TemplateAttendanceDateBinding card = DataBindingUtil.inflate(
                        inflater, R.layout.template_attendance_date, tr, true
                );
                CardView cardView = (CardView) card.getRoot();

                // setting table heading
                if (i == 0) {
                    cardView.setCardElevation(10);
                    cardView.setCardBackgroundColor(getColor(R.color.crystal_blue));
                    card.attendanceDateText.setText(shortWeekDays[j + 1]);
                }
            }
            table.addView(tr);
        }

        // Default year as current year
        Calendar cal = Calendar.getInstance();
        binding.inpAttendanceYear.setText(String.valueOf(cal.get(Calendar.YEAR)));

        // Autocomplete text
        List<String> months = Arrays.asList(getResources().getStringArray(R.array.months));
        binding.inpAttendanceMonth.setAdapter(new ArrayAdapter<>(
                activity, android.R.layout.simple_list_item_1, months));
        binding.inpAttendanceMonth.setText(months.get(cal.get(Calendar.MONTH)));

        binding.btnAttendanceGet.setOnClickListener(view1 -> {
            fetchAttendanceSheet();
            String y = CustomUtil.validateField(binding.attendanceYear, "year");
            String m = CustomUtil.validateField(binding.attendanceMonth, "text");
            List<String> mArray = Arrays.asList(getResources().getStringArray(R.array.months));

            if (y == null || m == null || !mArray.contains(m)) return;
            int year = Integer.parseInt(y);
            int month = mArray.indexOf(m);

            // call function to draw table
            setTable(table, month, year);
        });
    }

    // fill table with data
    public void setTable(TableLayout table, Integer month, Integer year) {
        // get Set of absent dates, and sessions from database for month
        Set<Integer> absentDateSet = new HashSet<>(getDatesInRange(absentList, year, month));
        Set<Integer> sessionDateSet = new HashSet<>(getDatesInRange(sessionList, year, month));

        Calendar calInstance = Calendar.getInstance();
        int currYear = calInstance.get(Calendar.YEAR);
        int currMonth = calInstance.get(Calendar.MONTH);

        // Check today is present in the selected view
        int today = (year == currYear && month == currMonth) ? calInstance.get(Calendar.DATE) : -1;

        // making custom Instance
        Calendar customInstance = Calendar.getInstance();
        customInstance.set(calInstance.get(Calendar.YEAR), (month < 0) ? currMonth : month,
                1, 0, 0, 0);
        customInstance.set(Calendar.MILLISECOND, 0);

        int maxDate = customInstance.getActualMaximum(Calendar.DATE);
        int startSpace = customInstance.get(Calendar.DAY_OF_WEEK) - 1;
        int dateCounter = 1;


        // Colors
        int black = getColor(R.color.black);
        int sessionColor = getColor(R.color.hunter_green);
        int absentColor = getColor(R.color.candy_red);
        int todayColor = getColor(R.color.moonstone);
        int white = getColor(R.color.white);

        // Looping through the table
        for (int i = 1; i < table.getChildCount(); i++) {
            TableRow row = (TableRow) table.getChildAt(i);
            row.setVisibility(View.GONE);
            for (int j = 0; j < row.getChildCount(); j++) {
                String date;

                // getting textView
                CardView card = (CardView) row.getChildAt(j);
                TextView text = (TextView) card.getChildAt(0);

                // per date logic
                card.setCardBackgroundColor(white);
                if ((i == 1 && j < startSpace) || (dateCounter > maxDate)) {
                    // no date in element
                    date = " ";
                } else {
                    // there is a date in the element
                    text.setTextColor(black);

                    // if first element in row, set row as visible
                    if (j == 0 || j == startSpace) row.setVisibility(View.VISIBLE);

                    // set color if the date is today
                    if (dateCounter == today) {
                        card.setCardBackgroundColor(todayColor);
                        text.setTextColor(white);
                    }
                    // set color if session occurred on date
                    if (sessionDateSet.contains(dateCounter)) {
                        card.setCardBackgroundColor(sessionColor);
                        text.setTextColor(white);
                    }

                    // set color if the student is absent on date
                    if (absentDateSet.contains(dateCounter)) {
                        card.setCardBackgroundColor(absentColor);
                        text.setTextColor(black);
                    }
                    date = String.valueOf(dateCounter++);
                }
                text.setText(date);

            }
        }
    }

    private void fetchAttendanceSheet() {
        selectedClass.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                if (!snapshot.exists()) return;
                ClassHelperClass currentClass = snapshot.getValue(ClassHelperClass.class);
                if (currentClass == null) return;
                if (absentList == null || !absentList.isEmpty())
                    absentList = new ArrayList<>();

                if (sessionList == null || !sessionList.isEmpty())
                    sessionList = new ArrayList<>();

                for (Map.Entry<String, Boolean> session : currentClass.sessions.entrySet()) {
                    String date = session.getKey();
                    sessionList.add(Long.parseLong(date));
                    String attendanceKey = CustomUtil.SHA1(userId + date);
                    AttendanceHelperClass record = currentClass.attendance.get(attendanceKey);
                    if (record != null) {
                        absentList.add(record.absent_date);
                    }
                }

                // Set progress bar
                int sessionCount = sessionList.size();
                int percent = sessionCount - absentList.size();
                setProgress(percent, sessionCount);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    private List<Integer> getDatesInRange(List<Long> dateList, int year, int month) {
        if (dateList.isEmpty()) {
            fetchAttendanceSheet();
            return Collections.emptyList();
        }

        // Set an instance to given year and month
        Calendar cal = Calendar.getInstance();
        cal.set(year, month, 1, 0, 0, 0);
        cal.set(Calendar.MILLISECOND, 0);

        // Get both bounds for filtering the absent list
        long minTime = cal.getTimeInMillis();
        cal.add(Calendar.MONTH, 1);
        long maxTime = cal.getTimeInMillis();

        // Filter absentList to get dates of selected range
        List<Integer> datesInRange = new ArrayList<>();
        for (long date : dateList) {
            if (minTime <= date && date < maxTime) {
                cal.setTimeInMillis(date);
                datesInRange.add(cal.get(Calendar.DATE));
            }
        }
        return datesInRange;
    }

    private void setProgress(int present, int sessionCount) {
        int progress;
        TextView detailed = binding.attendancePercentDetailed;

        if (sessionCount == 0 || present < 0) {
            progress = 100;
        } else {
            progress = 100 * present / sessionCount;
            String detail = "Session count: " + sessionCount +
                    "\nPresent count: " + present +
                    "\nAbsent  count: " + (sessionCount - present);
            detailed.setText(detail);
            binding.attendancePercentText.setOnClickListener(view -> {
                if (detailed.getVisibility() == View.VISIBLE) {
                    detailed.setVisibility(View.GONE);
                } else {
                    detailed.setVisibility(View.VISIBLE);
                }
            });
        }
        binding.attendanceProgressBar.setProgress(progress);
        binding.attendancePercentText.setText(String.format("%s%%", progress));
        binding.attendancePercentText.bringToFront();
    }

    private int getColor(int resColor) {
        return ResourcesCompat.getColor(getResources(), resColor, null);
    }
}

File: app/src/main/java/com/imperium/academio/ui/fragment/AttendanceStudent.kt
package com.imperium.academio.ui.fragment

import android.app.Activity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment
import com.google.firebase.database.*
import com.imperium.academio.CustomUtil.SHA1
import com.imperium.academio.CustomUtil.validateField
import com.imperium.academio.R
import com.imperium.academio.databinding.FragmentAttendanceStudentBinding
import com.imperium.academio.databinding.TemplateAttendanceDateBinding
import com.imperium.academio.fireclass.ClassHelperClass
import java.text.DateFormatSymbols
import java.util.*
import kotlin.collections.ArrayList

class AttendanceStudent : Fragment() {
    private lateinit var binding: FragmentAttendanceStudentBinding
    private lateinit var selectedClass: DatabaseReference
    private var classId: String? = null
    var userId: String? = null
    var absentList: MutableList<Long> = ArrayList()
    var sessionList: MutableList<Long> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val bundle = arguments
        if (bundle != null) {
            classId = bundle.getString("classId")
            userId = bundle.getString("userId")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        super.onCreateView(inflater, container, savedInstanceState)
        binding = FragmentAttendanceStudentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val activity: Activity? = activity
        if (activity == null || classId == null || userId == null) return

        selectedClass = FirebaseDatabase.getInstance().getReference("class/$classId")
        fetchAttendanceSheet()

        // Temporarily fill progress bar
        setProgress(0, 0)

        // Table creation
        val table = binding.attendanceStudentTable
        val shortWeekDays = DateFormatSymbols.getInstance().shortWeekdays

        for (i in 0..6) {
            // For row
            val tr = TableRow(activity)
            for (j in 0..6) {
                // For Element
                val card = TemplateAttendanceDateBinding.inflate(layoutInflater)
                val cardView = card.root as CardView

                // setting table heading
                if (i == 0) {
                    cardView.cardElevation = 10f
                    cardView.setCardBackgroundColor(getColor(R.color.crystal_blue))
                    card.attendanceDateText.text = shortWeekDays[j + 1]
                }
                tr.addView(cardView)
            }
            table.addView(tr)
        }

        table.isStretchAllColumns = true
        table.bringToFront()

        // Default year as current year
        val cal = Calendar.getInstance()
        val months = listOf(*resources.getStringArray(R.array.months))

        binding.run {
            inpAttendanceYear.setText(cal[Calendar.YEAR].toString())

            // Autocomplete text
            inpAttendanceMonth.setAdapter(ArrayAdapter(
                    activity, android.R.layout.simple_list_item_1, months))
            inpAttendanceMonth.setText(months[cal[Calendar.MONTH]])
            btnAttendanceGet.setOnClickListener {
                val y = validateField(binding.attendanceYear, "year")
                val m = validateField(binding.attendanceMonth, "text")
                val mArray = listOf(*resources.getStringArray(R.array.months))
                if (y == null || m == null || !mArray.contains(m)) return@setOnClickListener
                val year = y.toInt()
                val month = mArray.indexOf(m)

                // call function to draw table
                setTable(table, month, year)
            }
        }
    }

    // fill table with data
    private fun setTable(table: TableLayout, month: Int, year: Int) {
        // get Set of absent dates, and sessions from database for month
        val absentDateSet: Set<Int> = HashSet(getDatesInRange(absentList, year, month))
        val sessionDateSet: Set<Int> = HashSet(getDatesInRange(sessionList, year, month))
        val calInstance = Calendar.getInstance()
        val currYear = calInstance[Calendar.YEAR]
        val currMonth = calInstance[Calendar.MONTH]

        // Check today is present in the selected view
        val today = if (year == currYear && month == currMonth) calInstance[Calendar.DATE] else -1

        // making custom Instance
        val customInstance = Calendar.getInstance()
        customInstance[calInstance[Calendar.YEAR], if (month < 0) currMonth else month, 1, 0, 0] = 0
        customInstance[Calendar.MILLISECOND] = 0
        val maxDate = customInstance.getActualMaximum(Calendar.DATE)
        val startSpace = customInstance[Calendar.DAY_OF_WEEK] - 1
        var dateCounter = 1


        // Colors
        val black = getColor(R.color.black)
        val sessionColor = getColor(R.color.hunter_green)
        val absentColor = getColor(R.color.candy_red)
        val todayColor = getColor(R.color.moonstone)
        val white = getColor(R.color.white)

        // Looping through the table
        for (i in 1 until table.childCount) {
            val row = table.getChildAt(i) as TableRow
            row.visibility = View.GONE
            for (j in 0 until row.childCount) {
                var date: String

                // getting textView
                val card = row.getChildAt(j) as CardView
                val text = card.getChildAt(0) as TextView

                // per date logic
                card.setCardBackgroundColor(white)
                if (i == 1 && j < startSpace || dateCounter > maxDate) {
                    // no date in element
                    date = " "
                } else {
                    // there is a date in the element

                    // if first element in row, set row as visible
                    if (j == 0 || j == startSpace)
                        row.visibility = View.VISIBLE

                    text.setTextColor(black)
                    when {
                        // set color if the student is absent on date
                        absentDateSet.contains(dateCounter) -> {
                            card.setCardBackgroundColor(absentColor)
                        }

                        // set color if session occurred on date
                        sessionDateSet.contains(dateCounter) -> {
                            card.setCardBackgroundColor(sessionColor)
                            text.setTextColor(white)
                        }

                        // set color if the date is today
                        dateCounter == today -> {
                            card.setCardBackgroundColor(todayColor)
                            text.setTextColor(white)
                        }
                    }

                    date = dateCounter++.toString()
                }
                text.text = date
            }
        }
    }

    private fun fetchAttendanceSheet() {
        selectedClass.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (!snapshot.exists()) return
                val currentClass = snapshot.getValue(ClassHelperClass::class.java) ?: return
                if (absentList.isNotEmpty()) absentList = ArrayList()
                if (sessionList.isNotEmpty()) sessionList = ArrayList()
                for ((date) in currentClass.sessions) {
                    sessionList.add(date.toLong())
                    val attendanceKey = SHA1(userId + date)
                    val record = currentClass.attendance[attendanceKey]
                    if (record != null) {
                        absentList.add(record.absent_date)
                    }
                }

                // Set progress bar
                val sessionCount = sessionList.size
                val percent = sessionCount - absentList.size
                setProgress(percent, sessionCount)
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    private fun getDatesInRange(dateList: List<Long>?, year: Int, month: Int): List<Int> {
        if (dateList!!.isEmpty()) {
            fetchAttendanceSheet()
            return emptyList()
        }

        // Set an instance to given year and month
        val cal = Calendar.getInstance()
        cal[year, month, 1, 0, 0] = 0
        cal[Calendar.MILLISECOND] = 0

        // Get both bounds for filtering the absent list
        val minTime = cal.timeInMillis
        cal.add(Calendar.MONTH, 1)
        val maxTime = cal.timeInMillis

        // Filter absentList to get dates of selected range
        val datesInRange: MutableList<Int> = ArrayList()
        for (date in dateList) {
            if (date in minTime until maxTime) {
                cal.timeInMillis = date
                datesInRange.add(cal[Calendar.DATE])
            }
        }
        return datesInRange
    }

    private fun setProgress(present: Int, sessionCount: Int) {
        val progress: Int
        val detailed = binding.attendancePercentDetailed
        if (sessionCount == 0 || present < 0) {
            progress = 100
        } else {
            progress = 100 * present / sessionCount
            val detail = """
                Session count: $sessionCount
                Present count: $present
                Absent  count: ${sessionCount - present}
                """.trimIndent()
            detailed.text = detail
            binding.attendancePercentText.setOnClickListener {
                if (detailed.visibility == View.VISIBLE) {
                    detailed.visibility = View.GONE
                } else {
                    detailed.visibility = View.VISIBLE
                }
            }
        }
        binding.apply {
            attendanceProgressBar.progress = progress
            attendancePercentText.text = String.format("%s%%", progress)
        }
    }

    private fun getColor(resColor: Int): Int {
        return ResourcesCompat.getColor(resources, resColor, null)
    }

    companion object {
        fun newInstance(args: Bundle?): AttendanceStudent {
            return AttendanceStudent().apply {
                arguments = args
            }
        }
    }
}